### ISSUE TITLE
region() behavior at the feature boundaries

### ISSUE THREAD

**Issue by littleblackfish** [daler/gffutils#129]
In my gff3, I have a feature that is seqid=x, start=a, end=b. By gff3 convention, both a and b are included. If I query for the single position a with `region(seqid=x, start=a, end=a)` I am not finding this feature, however if I query `region(seqid=x, start=a, end=a+1)` or  `region(seqid=x, start=a+1, end=a+1) ` I do find it. This seems inconsistent to me. Am I missing something? Is this behavior defined anywhere?

**Comment by daler** [comment]
Sorry for the late reply. I agree that what you describe sounds inconsistent, but I'm unable to reproduce this behavior. Do you have an example file I could take a look at?

Here's what I tried:

```
# a.gff
1	.	gene	100	110	.	.	.	.
```

```python
import gffutils
db = gffutils.create_db('a.gff', ':memory:')
list(db.region(seqid='1', start=100, end=110))
# [<Feature gene (1:100-110[.]) at 0x7f9798d2a390>]
```

There are considerably more involved examples in the [tests for `region`](https://github.com/daler/gffutils/blob/master/gffutils/test/test.py#L432), it would help if you could let me know if you see something there that doesn't make sense.

**Comment by Brunox13** [comment]
I encountered a similar issue when I was experimentally trying to determine whether the arguments of the `gffutils.FeatureDB.region()` function are **1-based** or **0-based**.

Specifically, as an example, I downloaded the reference from 10X Genomics:
```bash
wget http://cf.10xgenomics.com/supp/cell-exp/refdata-cellranger-GRCh38-3.0.0.tar.gz
gunzip refdata-cellranger-GRCh38-3.0.0.tar.gz
```
Here, the GTF file is located in `./refdata-cellranger-GRCh38-3.0.0/genes/genes.gtf`.

For the purposes of this testing, I am looking at the following feature in this GTF file:
`1	ensembl	start_codon	69091	69093 [...]`

Clearly, a start codon has 3 bases, and so the GTF file is 1-based, and the [start, end] interval is closed. Now, I ran the following pieces of code:

I created and connected to the database using the following python code:

```python
import gffutils

gffutils.create_db(
    refdata-cellranger-GRCh38-3.0.0/genes/genes.gtf,
    dbfn = gffdb.db,
    force = True,
    merge_strategy = 'create_unique',
    id_spec = ['ID', 'Name'],
    verbose = True
    )

db = gffutils.FeatureDB(gffdb.db, keep_order = True)
```
Then I ran the following simple tests:

### 1.
```python
# The feature should NOT be included either way, whether 0-based or 1-based
for ft in db.region(region = (1, 69087, 69090), featuretype = 'start_codon'):
    print('{}: {}-{}'.format(ft.id, ft.start, ft.end))
```
Output: [empty]

### 2.
```python
# The feature SHOULD be included either way
for ft in db.region(region = (1, 69088, 69091), featuretype = 'start_codon'):
    print('{}: {}-{}'.format(ft.id, ft.start, ft.end))
```
Output: **[empty]**

### 3.
```python
# Should be included either way
for ft in db.region(region = (1, 69089, 69092), featuretype = 'start_codon'):
    print('{}: {}-{}'.format(ft.id, ft.start, ft.end))
```
Output: `start_codon_2: 69091-69093`
*Note that this clearly reflects the 1-based notation of the GTF file. If the start/end properties were converted to 0-based, I'd expect the output to be `start_codon_2: 69090-69093` where the [start, end) interval is half-open.*

### 4.
```python
# Should be included either way
for ft in db.region(region = (1, 69090, 69093), featuretype = 'start_codon'):
    print('{}: {}-{}'.format(ft.id, ft.start, ft.end))
```
Output: `start_codon_2: 69091-69093`

### 5.
```python
# Should be included either way
for ft in db.region(region = (1, 69091, 69094), featuretype = 'start_codon'):
    print('{}: {}-{}'.format(ft.id, ft.start, ft.end))
```
Output: `start_codon_2: 69091-69093`

### 6.
```python
# Should be included either way
for ft in db.region(region = (1, 69092, 69095), featuretype = 'start_codon'):
    print('{}: {}-{}'.format(ft.id, ft.start, ft.end))
```
Output: `start_codon_2: 69091-69093`

### 7.
```python
# Should NOT be included if 0-based. SHOULD be included if 1-based.
for ft in db.region(region = (1, 69093, 69096), featuretype = 'start_codon'):
    print('{}: {}-{}'.format(ft.id, ft.start, ft.end))
```
Output: **[empty]**

You can see that these results are inconsistent with either 0-based or 1-based notation of the `region()` function:
- If it was 0-based, output # 2 should not be empty.
- If it was 1-based, output # 2 and # 7 should not be empty.

**Comment by Brunox13** [comment]
After some further testing, I have confirmed that the following rule seems to be in place:

Both the `region()` query and the GTF/GFF database features are **1-based**, however, the `region()` query will exclude the set boundaries. That is, while the database features have **closed intervals [a,b]**, like the GTF/GFF file itself, the `region()` queries have **open intervals (x,y)**.

Using the example GTF file from my comment above, the 'gene_1' feature is as follows:
`1       havana  gene    29554   31109 [...]`

Here are the results from my test queries:

_Testing for the lower boundary of the feature..._
`gffutils.FeatureDB.region(region = (1, 29000, 29553))` -> NOT found (expected)
`gffutils.FeatureDB.region(region = (1, 29000, 29554))` -> NOT found (error)
`gffutils.FeatureDB.region(region = (1, 29000, 29555))` -> FOUND (expected)
`gffutils.FeatureDB.region(region = (1, 29554, 29554))` -> NOT found (error)
`gffutils.FeatureDB.region(region = (1, 29555, 29555))` -> FOUND (expected)

_Testing for the upper boundary of the feature..._
`gffutils.FeatureDB.region(region = (1, 31108, 31200))` -> FOUND (expected)
`gffutils.FeatureDB.region(region = (1, 31109, 31200))` -> NOT found (error)
`gffutils.FeatureDB.region(region = (1, 31110, 31200))` -> NOT found (expected)
`gffutils.FeatureDB.region(region = (1, 31108, 31108))` -> FOUND (expected)
`gffutils.FeatureDB.region(region = (1, 31109, 31109))` -> NOT found (error)

I believe this behavior is neither expected, nor documented. Until this is fixed, I suggest using a **1-based interval (x-1, y+1)**  in the `gffutils.FeatureDB.region()` function to query for the 1-based [x,y] interval.

**Comment by littleblackfish** [comment]
This is exactly my experience. Thank you very much for characterizing the bug. I was working on something else by the time @daler replied and never got around to reproducing it.

**Comment by yl3** [comment]
This issue seems to be caused by the fact that [these lines](https://github.com/daler/gffutils/blob/master/gffutils/interface.py#L637-L638) in `gffutils.FeatureDB.region()` involve strict inequalities. The boundary base pairs get included correctly, if the lines are changed into

    start_op = '<='
    end_op = '>='

Could someone else if this change is correct?

**Comment by husensofteng** [comment]
I wonder if there is any update on this issue? I am using `v0.9` and have a similar problem as shown below.

`GTF` entry:

> 11	ensembl_havana	exon	803174	803587	.	-	.	info

intervals that should overlap the last nucleotide of the exon above:

> 11:803587-803588 
> 11:803587-803587

but none of them return any overlap.

**Comment by husensofteng** [comment]
> This issue seems to be caused by the fact that [these lines](https://github.com/daler/gffutils/blob/master/gffutils/interface.py#L637-L638) in `gffutils.FeatureDB.region()` involve strict inequalities. The boundary base pairs get included correctly, if the lines are changed into
> 
> ```
> start_op = '<='
> end_op = '>='
> ```
> 
> Could someone else check if this change is correct?

this fix does not seem to work. at least it doesn't report any overlap for the example above.

**Comment by husensofteng** [comment]
I have made [pull request 161](https://github.com/daler/gffutils/pull/162) addressing this issue.

**Comment by daler** [comment]
Thanks @husensofteng, this is now addressed in v0.11.
